flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java

```diff
@@ -51,39 +48,52 @@
  * @param <TXN> Transaction to store all of the information required to handle a transaction (must be Serializable)
  */
 @PublicEvolving
-public abstract class TwoPhaseCommitSinkFunction<IN, TXN extends Serializable>
+public abstract class TwoPhaseCommitSinkFunction<IN, TXN>
 		extends RichSinkFunction<IN>
 		implements CheckpointedFunction, CheckpointListener {
 
 	private static final Logger LOG = LoggerFactory.getLogger(TwoPhaseCommitSinkFunction.class);
 
-	protected final ListStateDescriptor<TransactionAndCheckpoint<TXN>> pendingCommitTransactionsDescriptor;
+	protected final ListStateDescriptor<Map<Long, TXN>> pendingCommitTransactionsDescriptor;
 	protected final ListStateDescriptor<TXN> pendingTransactionsDescriptor;
 
-	protected final List<TransactionAndCheckpoint<TXN>> pendingCommitTransactions = new ArrayList<>();
+	protected final LinkedHashMap<Long, TXN> pendingCommitTransactions = new LinkedHashMap<>();
 
 	@Nullable
 	protected TXN currentTransaction;
 	protected ListState<TXN> pendingTransactionsState;
-	protected ListState<TransactionAndCheckpoint<TXN>> pendingCommitTransactionsState;
-
-	public TwoPhaseCommitSinkFunction(Class<TXN> txnClass) {
-		this(
-			TypeInformation.of(txnClass),
-			TypeInformation.of(new TypeHint<TransactionAndCheckpoint<TXN>>() {}));
-	}
+	protected ListState<Map<Long, TXN>> pendingCommitTransactionsState;
```

Comment:
Whether transactions can be redistributed depends on the system that we communicate with. For Kafka 0.11 I'm not sure. Now when I think about it a little bit more, I will change it to `ListState<List<Tuple2<Long, TXN>>>`, so that we can guarantee that transactions will be recovered in the same order they were created.